containers-tests/tests/tree-properties.hs

```diff
@@ -101,3 +116,23 @@ prop_monadFix_ls val ta ti =
     f :: (Int -> Int) -> Int -> Tree (Int -> Int)
     f q y = let t = apply ti y
             in fmap (\w -> fact w q) t
+
+prop_toList :: Tree Int -> Property
+prop_toList t = toList t === foldr (:) [] t
+
+prop_foldMap :: Tree Int -> Property
+prop_foldMap t = foldMap (:[]) t === toList t
+
+prop_foldl' :: Tree Int -> Property
+prop_foldl' t = foldl' (flip (:)) [] t === reverse (toList t)
+
+prop_foldr1 :: Tree Int -> Property
+prop_foldr1 t = foldr1 (++) (fmap (:[]) t) === toList t
+
+prop_foldl1 :: Tree Int -> Property
+prop_foldl1 t = foldl1 (++) (fmap (:[]) t) === toList t
+
+prop_foldr_infinite :: Int -> Property
+prop_foldr_infinite n = take n (foldr (:) [] t) === [1..n]
+  where
+    t = Node 1 [Node x [] | x <- [2..]]
```

Comment:
This test is really just about termination, right? Anything else it has to say will be captured by finite tests. Wouldn't it be better to use something like this to get more infinite tree shapes?

```haskell
prop_foldr_infinite n = forAll genny (\genny -> length (take n (foldr (:) [] t)) === n)
  where
    genny :: Gen (Tree ())
    genny = Node () <$> oneof [listOf genny, infiniteListOf genny]
```

Reply:
Sure, done. This generator can generate a finite tree though (`listOf genny` can be null), so I added an infinite of these at the top level.

Reply:
Generating a finite tree sometimes is fine. We're going to be generating a *lot* of trees. I'd keep it simple.

Reply:
Oh, but I forgot that it should therefore be `length ... <= n`!

Reply:
Oh that's what you meant, done.